### PR TITLE
[SDK-365] Internal Limits

### PR DESCRIPTION
**Pull request by turtledreams**
SDK-365

Internal limits implemented

### CODE REVIEW

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Move the truncation functions to the common module

**Comment by ArtursKadikis** [comment]
+ there is a merge conflict

**Comment by turtledreams** [comment]
image

Functions were moved to common module and everything working as original

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
"countly-common" needs to be reworked to an object. 
The "require" path should be shortened.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Add a few tests that would verify this functionality

**Comment by turtledreams** [comment]
image

**Comment by turtledreams** [comment]
image

File: test/test-internal-limit.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        var errorLog = "test";
+        var newStr = cc.truncateSingleValue(string, limit, errorLog);
+        assert.equal(newStr.length, 3);
+    });
+    it("Check if the number is truncated", function() {
+        var number = 123456789;
+        var limit = 3;
+        var errorLog = "test";
+        var newStr = cc.truncateSingleValue(number, limit, errorLog);
+        assert.equal(newStr.length, 3);
```
Comment (ArtursKadikis) [inline test/test-internal-limit.js:19]: verify the resulting string and not just it's length. Theoretically the length might match but the content might be garbage

File: test/test-internal-limit.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+var cc = require("../lib/countly-common");
+
+// Unit tests for truncateSingleValue:
+describe("Internal limits single value truncation testing suite", function() {
+    it("Check if the string is truncated", function() {
+        var string = "123456789";
+        var limit = 3;
+        var errorLog = "test";
+        var newStr = cc.truncateSingleValue(string, limit, errorLog);
+        assert.equal(newStr.length, 3);
```
Comment (ArtursKadikis) [inline test/test-internal-limit.js:12]: verify the resulting content

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
your tests currently are missing usecases that validate the result at the request queue of event queue level.

The expected loop was:
1) set limits
2) init SDK
3) use some SDK feature while exceeding limits
4) verify in RQ or EQ that values are truncated

**Comment by ArtursKadikis** [comment]
Merge conflicts

**Comment by turtledreams** [comment]
image

Do we need changelog/docs entries for halt?

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Move specific functions to the test function.
Update "halt" comment.

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**